Lib/traceback.py

```diff
@@ -586,8 +587,11 @@ def _format_syntax_error(self, stype):
         """Format SyntaxError exceptions (internal helper)."""
         # Show exactly where the problem was found.
         filename = self.filename or "<string>"
-        lineno = str(self.lineno) or '?'
-        yield '  File "{}", line {}\n'.format(filename, lineno)
+        filename_suffix = ''
+        if self.lineno is not None:
+            yield '  File "{}", line {}\n'.format(filename, self.lineno)
+        elif self.filename is not None:
+            filename_suffix = ' ({})'.format(self.filename)
```

Comment:
Since `filename` is only used in the first "if" block, but `self.filename` is used later, I'd move its definition into the block, or just remove it and use `self.filename or "<string>"` inline.